File: docs/chapters/02-design-patterns/java/CircuitBreakerFraudClient.java
// Scenario: optional fraud score enrichment on checkout (non-blocking path)
// Demonstrates: Circuit breaker — fail fast when dependency is unhealthy
// Trade-off: degraded checkout vs hard fail; tune with timeouts and metrics

import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

final class CircuitBreakerFraudClient {
    enum State { CLOSED, OPEN, HALF_OPEN }
    private volatile State state = State.CLOSED;
    private final AtomicInteger failures = new AtomicInteger();
    private final int threshold = 5;
    private final FraudApi api;

    CircuitBreakerFraudClient(FraudApi api) { this.api = api; }

    Optional<Integer> score(String orderId) {
        if (state == State.OPEN) return Optional.empty();
        try {
            int s = api.fetchScore(orderId);
            failures.set(0);
            state = State.CLOSED;
            return Optional.of(s);
        } catch (Exception e) {
            if (failures.incrementAndGet() >= threshold) state = State.OPEN;
            return Optional.empty();
        }
    }

    interface FraudApi { int fetchScore(String orderId); }
}

File: chapters/02-design-patterns/go/circuit_breaker_fraud_client.go
// Scenario: optional fraud score enrichment on checkout
// Demonstrates: Circuit breaker sketch — production use gobreaker/resilience4j
// Trade-off: library handles half-open; here shows state + fail-fast intent

package main

import (
	"errors"
	"sync/atomic"
)

type FraudAPI interface {
	FetchScore(orderID string) (int, error)
}

type CircuitBreakerFraudClient struct {
	api       FraudAPI
	failures  atomic.Int32
	threshold int32
	open      atomic.Bool
}

func (c *CircuitBreakerFraudClient) Score(orderID string) (int, bool) {
	if c.open.Load() {
		return 0, false
	}
	score, err := c.api.FetchScore(orderID)
	if err != nil {
		if c.failures.Add(1) >= c.threshold {
			c.open.Store(true)
		}
		return 0, false
	}
	c.failures.Store(0)
	return score, true
}

var ErrFraudUnavailable = errors.New("fraud unavailable")
